source/host_resolver.c

```diff
@@ -1473,8 +1475,9 @@ static int default_resolve_host(
         if (aws_array_list_length(&callback_address_list)) {
             res(resolver, host_name, AWS_OP_SUCCESS, &callback_address_list, user_data);
         } else {
-            res(resolver, host_name, aws_last_error(), NULL, user_data);
-            result = AWS_OP_ERR;
+            int last_error = aws_last_error();
+            int error_code = (last_error != AWS_ERROR_SUCCESS) ? last_error : AWS_IO_DNS_QUERY_FAILED;
+            return error_code;
```

Comment:
our `int` functions return either AWS_OP_SUCCESS or AWS_OP_ERROR (and then users call aws_last_error() to get the actual error code). We never return the actual error code itself from an `int`-returning function

Reply:
also ... how are we not leaking memory by suddenly doing a return here?

isn't there an aws_array_list_clean_up() that still needs to happen?

Reply:
Thanks, I have fixed it.